doc/dev/adr/0025-rpc-streaming.md

```diff
@@ -0,0 +1,399 @@
+
+Since sending a cancellation request may fail (message expiry on broker side), the SDK API design should allow for the user to repeatedly call "cancel" and should return successfully once the other party has responded appropriately. 
+
+Additionally, cancellation requests may include user properties. This allows users to provide additional context on why the cancellation is happening.
+
+#### .NET API design
+
+The proposed cancellation support would come from the return type on the invoker side and the provided type on the executor side:
+
+```csharp
+public interface IStreamContext<T>
+    where T : class
+{
+    /// <summary>
+    /// The asynchronously readable entries in the stream.
+    /// </summary>
+    IAsyncEnumerable<T> Entries { get; set; }
+
+    /// <summary>
+    /// Cancel this received RPC streaming request.
+    /// </summary>
+    /// <param name="userProperties">The optional user properties to include</param>
+    /// <param name="cancellationToken">Cancellation token for this cancellation request</param>
+    /// <remarks>
+    /// This method may be called by the streaming executor at any time. For instance, if the request stream
+    /// stalls unexpectedly, the executor can call this method to notify the invoker to stop sending requests.
+    /// Additionally, the executor can call this method if its response stream has stalled unexpectedly.
+    /// </remarks>
+    Task CancelAsync(Dictionary<string, string>? userProperties = null, CancellationToken cancellationToken = default);
+
+    /// <summary>
+    /// The token that tracks if the streaming exchange has been cancelled by the other party and/or timed out.
+    /// </summary>
+    /// <remarks>
+    /// For instance, if the invoker side cancels the streaming exchange, the executor side callback's <see cref="IStreamContext{T}.CancellationToken"/>
+    /// will be triggered. If the executor side cancels the streaming exchange, the invoker side's returned <see cref="IStreamContext{T}.CancellationToken"/>
+    /// will be triggered.
+    ///
+    /// To see if this was triggered because the stream exchange was cancelled, see <see cref="IsCanceled"/>. To see if it was triggered because
+    /// the stream exchange timed out, see <see cref="HasTimedOut"/>.
+    /// </remarks>
+    CancellationToken CancellationToken { get; }
+
+    /// <summary>
+    /// Get the user properties associated with a cancellation request started with <see cref="CancelAsync(Dictionary{string, string}?, CancellationToken)"/>.
+    /// </summary>
+    /// <returns>The user properties associated with a cancellation request</returns>
+    /// <remarks>
+    /// If the stream has not been cancelled, this will return null. If the stream has been cancelled, but no user properties were
+    /// provided in that cancellation request, this will return null.
+    /// </remarks>
+    Dictionary<string, string>? GetCancellationRequestUserProperties();
```

Comment:
why not have these as some information provided with the cancellation? (Or is this just a dotnet limitation)